force-app/main/default/classes/SOQL.cls

```diff
@@ -271,91 +338,151 @@ public with sharing class SOQL {
      */
     public static List<Contact> getAccountsWithContactsByState(Id accountId) {
         // Query the Account
-
+        String billingState = [SELECT BillingState FROM Account Where ID = :accountId].BillingState;
         // Store the Account's BillingState
-
-        // Query the Contacts
-        return null; // Replace null with the variable you used to store the result
-    }
-
-    /**
+        List<Contact> contactList = [SELECT Id, Name, MailingState FROM Contact WHERE MailingState = :billingState];
+        return contactList;     
+    }   
+ 
+    /*
      * Question 22
      * Extracts email domains from them leads, then finds Accounts with matching website domains
      * If Lead Email = test@test.com return account with website = www.test.com
      * Example: findContactsByLeadEmailDomains([set of lead ids]) should return a list of Accounts with matching email domains
      * @param A set of Lead Ids.
      * @return A list of Accounts with matching email domains.
      */
-    public static List<Account> findContactsByLeadEmailDomains(Set<Id> leadIds) {
-        // Query the Leads
-
-        // Create a set of email domains from the Leads
-
-        // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
-        
-        // Query the Accounts
-        return null; // Replace null with the variable you used to store the result
+     
+     public static List<Account> findContactsByLeadEmailDomains(Set<Id> leadIds) {
+    Set<String> emailDomains = new Set<String>();
+    
+    // Query the Leads
+    List<Lead> leadList = [SELECT Id, Email FROM Lead WHERE Id IN :leadIds];
+    
+    // Extract email domains from Leads' emails and add to the set
+    for (Lead lead : leadList) {
+        if (lead.Email != null) {
+            String[] parts = lead.Email.split('@');
+            if (parts.size() == 2) {
+                emailDomains.add(parts[1].toLowerCase()); // Convert to lower case for case insensitivity
+            }
+        }
     }
+    
+    // Query the Accounts with matching website domains
+    List<Account> matchingAccounts = [SELECT Id, Website FROM Account WHERE Website != null];
+    
+    // Filter the matching accounts based on email domains
+    List<Account> filteredAccounts = new List<Account>();
+    for (Account acc : matchingAccounts) {
+        for (String domain : emailDomains) {
+            if (acc.Website.containsIgnoreCase(domain)) {
+                filteredAccounts.add(acc);
+                break; // Once found, break and move to the next account
+            }
+        }
+    }
```

Comment:
This is a good introductory way of completing this.
How can you optimize this to reduce the nested loops?